gapis/resolve/dependencygraph2/builder.go

```diff
@@ -0,0 +1,303 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dependencygraph2
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/google/gapid/core/app/benchmark"
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/math/interval"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/config"
+	"github.com/google/gapid/gapis/memory"
+)
+
+var (
+	dependencyGraphBuilderCounter = benchmark.Duration("DependencyGraph.Builder")
+)
+
+// DependencyGraphBuilder builds a dependency graph from the effects in a capture
+type DependencyGraphBuilder interface {
+	Graph() DependencyGraph
+	api.StateWatcher
+}
+
+// The data needed to build a dependency graph by iterating through the commands in a trace
+type dependencyGraphBuilder struct {
+	// graph is the dependency graph being constructed
+	graph *dependencyGraph
+
+	// currentCmdID is the CmdID of the command currently being processed
+	currentCmdID api.CmdID
+
+	// currentCmd is the command currently being processed
+	currentCmd api.Cmd
+
+	// currentNodeID is the NodeID of the dependency graph node corresponding to the currentCmdID
+	currentNodeID NodeID
+
+	// currentDependencyExists[i] is true if currentNodeID depends on node i
```

Comment:
Why don't we use `map[NodeID]struct{}` to store current dependencies, it seems such an unordered_set can replace `currentDependencyExists` and `currentDependencies` and also avoid the sorting and panic hassle.

Reply:
Using the two slices significantly improved performance over a map. The dependency graph can easily have billions of dependencies, and, when there is a dependency between two nodes, there are often 10s or 1000s of different reads/writes that give the same dependency. Each of these redundant dependencies result in an array lookup in `currentDependencyExists`, rather than an attempted map insertion.

I don't think I kept the profiling data around (this was when I was doing a lot of experimentation with this code), but as I recall the overall improvement in the time to construct large dependency graphs improved by 10-20%.